xla/hlo/transforms/simplifiers/sub_byte_collective_normalization.cc

```diff
@@ -0,0 +1,208 @@
+  return input->parent()->AddInstruction(
+      HloInstruction::CreateBitcast(shape, convert));
+}
+
+bool CanBeRepresentedAs(const Shape& shape, const PrimitiveType casted_type) {
+  const int64_t ratio = primitive_util::BitWidth(casted_type) /
+                        primitive_util::BitWidth(shape.element_type());
+  return primitive_util::IsSubByteNonPredType(shape.element_type()) &&
+         ShapeUtil::LastDimIsMinorMost(shape) &&
+         shape.layout().element_size_in_bits() ==
+             primitive_util::BitWidth(shape.element_type()) &&
+         shape.dimensions().back() % ratio == 0;
+}
+
+class SubByteCollectiveNormalizationVisitor : public DfsHloRewriteVisitor {
+ public:
+  SubByteCollectiveNormalizationVisitor() = default;
+  absl::Status HandleAllGather(HloInstruction* hlo) override;
+  absl::Status HandleAllToAll(HloInstruction* hlo) override;
+  absl::Status HandleCollectiveBroadcast(HloInstruction* hlo) override;
+  absl::Status HandleCollectivePermute(HloInstruction* hlo) override;
+
+ private:
+  absl::Status ProcessCollectiveInstruction(HloInstruction& hlo);
+  static constexpr PrimitiveType casted_type_ = S8;
+};
+
+absl::Status SubByteCollectiveNormalizationVisitor::HandleAllGather(
+    HloInstruction* hlo) {
+  return ProcessCollectiveInstruction(*hlo);
+}
+
+absl::Status SubByteCollectiveNormalizationVisitor::HandleAllToAll(
+    HloInstruction* hlo) {
+  if (hlo->operand_count() != 1) {
+    // Variadic ones are not supported yet.
+    return absl::OkStatus();
+  }
+  if (!CanBeRepresentedAs(hlo->operand(0)->shape(), casted_type_)) {
+    return absl::OkStatus();
+  }
+
+  const int64_t ratio = primitive_util::BitWidth(casted_type_) /
+                        primitive_util::BitWidth(hlo->shape().element_type());
+  const auto* all_to_all = Cast<HloAllToAllInstruction>(hlo);
+  if (all_to_all->split_dimension()) {
+    TF_ASSIGN_OR_RETURN(const CollectiveOpGroupMode group_mode,
+                        GetCollectiveOpGroupMode(all_to_all));
+    const int64_t split_dimension_size =
+        hlo->shape().dimensions(*all_to_all->split_dimension());
+    if (split_dimension_size %
+            (GetSubgroupSize(all_to_all, group_mode) * ratio) !=
+        0) {
+      return absl::OkStatus();
+    }
+  }
+  return ProcessCollectiveInstruction(*hlo);
+}
+
+absl::Status SubByteCollectiveNormalizationVisitor::HandleCollectiveBroadcast(
+    HloInstruction* hlo) {
+  return ProcessCollectiveInstruction(*hlo);
+}
+
+absl::Status SubByteCollectiveNormalizationVisitor::HandleCollectivePermute(
+    HloInstruction* hlo) {
+  return ProcessCollectiveInstruction(*hlo);
+}
+
+absl::Status
+SubByteCollectiveNormalizationVisitor::ProcessCollectiveInstruction(
+    HloInstruction& hlo) {
+  if (hlo.operand_count() != 1) {
+    // Variadic ones are not supported yet.
+    return absl::OkStatus();
+  }
+  if (!CanBeRepresentedAs(hlo.operand(0)->shape(), casted_type_)) {
+    return absl::OkStatus();
+  }
```

Comment:
Can we factor out this logic instead of duplicating it? I.e.
```c++
bool SubByteCollectiveNormalizationVisitor::ShouldProcessInstruction(const HloInstruction& hlo) const {
  return hlo.operand_count() == 1 && CanBeRepresentedAs(hlo.operand(0)->shape(), casted_type_);
}
```